ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java

```diff
@@ -0,0 +1,201 @@
+        metricsSystem.createLabelledCounter(
+            BesuMetricCategory.PEERS,
+            "PeerTaskExecutor:InvalidResponseCounter",
+            "Counter of the number of invalid responses received",
+            "className");
+    internalExceptionCounter =
+        metricsSystem.createLabelledCounter(
+            BesuMetricCategory.PEERS,
+            "PeerTaskExecutor:InternalExceptionCounter",
+            "Counter of the number of internal exceptions occurred",
+            "className");
+    inflightRequestGauge =
+        metricsSystem.createLabelledGauge(
+            BesuMetricCategory.PEERS,
+            "PeerTaskExecutor:InflightRequestGauge",
+            "Gauge of the number of inflight requests",
+            "className");
+    inflightRequestCountByClassName = new ConcurrentHashMap<>();
+  }
+
+  public <T> PeerTaskExecutorResult<T> execute(final PeerTask<T> peerTask) {
+    PeerTaskExecutorResult<T> executorResult;
+    int retriesRemaining = peerTask.getRetriesWithOtherPeer();
+    final Collection<EthPeer> usedEthPeers = new HashSet<>();
+    do {
+      Optional<EthPeer> peer =
+          peerSelector.getPeer(
+              (candidatePeer) ->
+                  peerTask.getPeerRequirementFilter().test(candidatePeer)
+                      && !usedEthPeers.contains(candidatePeer));
+      if (peer.isEmpty()) {
+        executorResult =
+            new PeerTaskExecutorResult<>(
+                Optional.empty(), PeerTaskExecutorResponseCode.NO_PEER_AVAILABLE);
+        continue;
+      }
+      usedEthPeers.add(peer.get());
+      executorResult = executeAgainstPeer(peerTask, peer.get());
+    } while (retriesRemaining-- > 0
+        && executorResult.responseCode() != PeerTaskExecutorResponseCode.SUCCESS);
+
+    return executorResult;
+  }
+
+  public <T> PeerTaskExecutorResult<T> executeAgainstPeer(
+      final PeerTask<T> peerTask, final EthPeer peer) {
+    String taskClassName = peerTask.getClass().getSimpleName();
+    AtomicInteger inflightRequestCountForThisTaskClass =
+        inflightRequestCountByClassName.computeIfAbsent(
+            taskClassName,
+            (k) -> {
+              AtomicInteger inflightRequests = new AtomicInteger(0);
+              inflightRequestGauge.labels(inflightRequests::get, taskClassName);
+              return inflightRequests;
+            });
+    MessageData requestMessageData = peerTask.getRequestMessage();
+    PeerTaskExecutorResult<T> executorResult;
+    int retriesRemaining = peerTask.getRetriesWithSamePeer();
+    do {
+      try {
+        T result;
+        try (final OperationTimer.TimingContext ignored =
+            requestTimer.labels(taskClassName).startTimer()) {
+          inflightRequestCountForThisTaskClass.incrementAndGet();
+
+          MessageData responseMessageData =
+              requestSender.sendRequest(peerTask.getSubProtocol(), requestMessageData, peer);
+
+          result = peerTask.parseResponse(responseMessageData);
```

Comment:
It might be a good idea to not do the parsing here. I don't think that would be hard to change. It would make the PeerTaskExecutor almost trivial (no peer handling, no retries). The tasks themselves would have to handle that, but I guess they have to handle the partial success anyways. 
Could we have an AbstractTask that does most of that work?
Maybe something we can discuss tomorrow?

Reply:
The whole point of the executor is handle retries and peer switching for a task. The task classes themselves don't actively do anything, they're more like a collection of methods that do small specific actions or supply values, like generating the request message. Changing this would be a pretty substantial change at this stage, and creating an AbstractPeerTask class is starting to sound extremely similar to the old system we're trying to move away from.